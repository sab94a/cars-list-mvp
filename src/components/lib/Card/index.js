// @flow

import React from 'react';
import cx from 'classnames';
import type { Card as Props} from 'components/lib/types';
import styles from './index.module.scss';

const Card = ({ image, title, description, footer, loading, elemType = 'div' }: Props) => {
    const rootClasses = cx(styles.root, { [styles.loading]: loading });
    const imageStyle = { backgroundImage: loading || !image ? null : `url(${ image })` }
    const Component = elemType
    return (
        <Component className={ rootClasses }>
            <div className={ styles.image } style={ imageStyle } />
            <div className={ styles.content }>
                <h3 className={ styles.title }>{ title }</h3>
                <p className={ styles.description }>{ description }</p>
                <div className={ styles.footer }>
                    { footer }
                </div>
            </div>
        </Component>
    );
};

export default React.memo<Props>(Card);
