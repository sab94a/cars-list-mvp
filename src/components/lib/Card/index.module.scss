@import 'components/lib/variables.scss';

.root {
    display: flex;
    width: 100%;
    padding: $offset-sm;
    border: $border-width-sm solid $light;
    margin-top:  $offset-md;

    &.loading {
        .image, .title, .description, .footer > * {
            color:  $light;
            background: $light;
        }

        .title {
            height: 26px;
        }

        .description {
            height: 14px;
        }

        .footer:after {
            content: '';
            @extend .footerLink;
            width: 65px;
            height: 14px;
            color:  $light;
            background: $light;
        }
    }
}

.image {
    width: 90px;
    background-position: center;
    background-size: contain;
    background-repeat: no-repeat;
}

.content {
    padding: 3px 0 3px $offset-lg;
    min-width: 245px;
}

.title {
    margin: 0;
    margin-bottom: 2px;
    padding-bottom: $offset-sm;
    line-height: 1;
    font-size: $font-size-md;
    font-weight: bold;
}

.description {
    margin: 0;
}

.footerLink {
    display: inline-block;
    vertical-align: middle;
    margin-right: $border-width;
    text-decoration: none;
    margin-top: $offset-sm;
}

.footer {
    & > * {
        @extend .footerLink;
    }
}
