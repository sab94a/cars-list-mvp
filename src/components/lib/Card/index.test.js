import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Card from './';

describe('<Card />', () => {
    let component;
    const image = './image';
    const title = 'title';
    const description = 'description';
    const footer = (<b>Footer</b>);

    const create = params => {
        component = mount(<Card { ...params } />);
    };

    it('Should render correctly default view', () => {
        create({ image, title, description, footer });

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should render correctly loading state', () => {
        create({ loading: true });

        expect(toJson(component)).toMatchSnapshot();
    });

    it('Should have correct image, title, description and footer', () => {
        create({ image, title, description, footer });

        const $image = component.find('.image');
        const $title = component.find('.title');
        const $description = component.find('.description');
        const $footer = component.find('.footer');

        expect($image.prop('style')).toEqual({ backgroundImage: `url(${image})`});
        expect($title.text()).toBe(title);
        expect($description.text()).toBe(description);
        expect($footer.contains(footer)).toBeTruthy();
    });

    it('Should render correct element type', () => {
        create({ loading: true, elemType: 'article' });

        expect(component.find('.root').type()).toBe('article')
    })
});
