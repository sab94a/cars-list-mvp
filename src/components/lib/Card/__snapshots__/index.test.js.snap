// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Should render correctly default view 1`] = `
<Memo(Card)
  description="description"
  footer={
    <b>
      Footer
    </b>
  }
  image="./image"
  title="title"
>
  <div
    className="root"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(./image)",
        }
      }
    />
    <div
      className="content"
    >
      <h3
        className="title"
      >
        title
      </h3>
      <p
        className="description"
      >
        description
      </p>
      <div
        className="footer"
      >
        <b>
          Footer
        </b>
      </div>
    </div>
  </div>
</Memo(Card)>
`;

exports[`<Card /> Should render correctly loading state 1`] = `
<Memo(Card)
  loading={true}
>
  <div
    className="root loading"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": null,
        }
      }
    />
    <div
      className="content"
    >
      <h3
        className="title"
      />
      <p
        className="description"
      />
      <div
        className="footer"
      />
    </div>
  </div>
</Memo(Card)>
`;
